File: SqlOrganize/Sql/EntityMetadata.cs
namespace SqlOrganize.Sql
{
    public class EntityMetadata
    {
        public Db db { get; set; }

        public string name { get; set; }
        
        public string alias { get; set; }

        public string? schema { get; set;  }

        public List<string> pk { get; set; } = new();
        public List<string> fieldNames => fields.Keys.ToList();
        public List<string> fk { get; set; } = new();

        protected List<Field> _ref; //ref

        protected List<Field> _oor; //one to one (fk in ref)

        protected List<Field> _om; //one to many

        /// <summary> Array dinamico para identificar a la entidad en un momento determinado </summary>
        /// <example> ["fecha_anio", "fecha_semestre","persona__numero_documento"] </example>
        public List<string> identifier { get; set; } = new();

        /// <summary> Valores por defecto para ordenamiento </summary>
        /// <example> ["field1"=>"asc","field2"=>"desc",...]; </example>
        public Dictionary<string, string> orderDefault { get; set; } = new();

        /// <summary> nombres de campos no administrables </summary>
        /// <remarks> los campos no insertables son aquellos que reciben se asignan directamente desde el servidor sql </remarks>
        public List<string> noAdmin { get; set; } = new();


        /// <summary> Valores principales </summary>
        public List<string> main { get; set; } = new();

        /// <summary> Valores unicos </summary>
        /// <remarks> Una entidad puede tener varios campos que determinen un valor único </remarks>
        /// <example> field1, field2, field3, ... </example>
        public List<string> unique { get; set; } = new();

        /// <summary> Valores no nulos </summary>
        public List<string> notNull { get; set; } = new();

        /// <summary> Valores unicos multiples </summary>
        /// <remarks> Cada juego de valores unicos multiples se define como una Lista </remarks>
        public List<List<string>> uniqueMultiple { get; set; } = new();

        public Dictionary<string, EntityTree> tree { get; set; } = new();

        public Dictionary<string, EntityRelation> relations { get; set; } = new();

        public Dictionary<string, EntityRef> oo { get; set; } = new();
        public Dictionary<string, EntityRef> om { get; set; } = new();


        public string schema_ => String.IsNullOrEmpty(schema) ? schema : "";
        public string schemaName => schema + name;
        public string schemaNameAlias => schema + name + " AS " + alias;

        /*
        Campo de identificacion
        *  Si existe un solo campo pk, entonces la pk sera el id. 
        *  Si existe al menos un campo unique not null, se toma como id.     
        *  Si existe multiples campos pk, se toman la concatenacion como id. 
        *  Si existe multiples campos uniqueMultiple, se toman la concatenacion como id. 
        */
        public List<string> id { get; set; }


        public Dictionary<string, Field> fields { get; set; } = new();

        protected List<Field> _Fields(List<string> fieldNames)
        {
            List<Field> fields = new();
            foreach (string fieldName in fieldNames)
                fields.Add(db.Field(name, fieldName));

            return fields;

        }

        /*
        fields no fk
        */
        public List<Field> Fields() => _Fields(fieldNames);

        /*
        fields many to one
        */
        public List<Field> FieldsFk() => _Fields(fk);

        public List<Field> FieldsRef()
        {

            if (_ref != null)
                return _ref;

            _ref = new();
            _oor = new();
            _om = new();

            foreach (var (entityName, entity) in db.entities)
            {
                foreach(var fieldName in entity.fk)
                {
                    Field field = db.Field(entityName, fieldName);
                    if (field.refEntityName!.Equals(name))
                    {
                        _ref.Add(field);
                        if (entity.unique.Contains(field.name))
                            _oor.Add(field);
                        else
                            _om.Add(field);
                    }
                }
            }

            return _ref;
        }

        public List<Field> FieldsOor()
        {

            if (_oor != null)
                return _oor;

            FieldsRef();

            return _oor!;
        }

        public List<Field> FieldsOm()
        {

            if (_om != null)
                return _om;

            FieldsRef();

            return _om!;
        }


    }
}

File: SqlOrganize/Sql/EntityMetadata.php
<?php

namespace SqlOrganize\Sql;

require_once MAIN_PATH . 'SqlOrganize/Utils/ValueTypesUtils.php';

use SqlOrganize\Utils\ValueTypesUtils;

class EntityMetadata
{
    public Db $db;
    public string $name;
    public string $alias;
    public ?string $schema = null;
    
    private ?string $_className = null;
    
    public static function getInstance($name, $alias){
        $em = new EntityMetadata();
        $em->name = $name;
        $em->alias = $alias;
        return $em;
    }

    public function getClassName(): string
    {
        return empty($this->_className) 
            ? ValueTypesUtils::toCamelCase($this->name)
            : $this->_className;
    }

    public function getClassNameWithNamespace(): string
    {
        $namespace = (!empty($this->db->config->namespace)) ? "\\".$this->db->config->namespace ."\\" : "";
        return $namespace . $this->getClassName();
    }
    
    public function setClassName(string $className): void
    {
        $this->_className = $className;
    }
    
    /** @var string[] */
    public array $pk = [];
    
    /** @var string[] */
    public array $fk = [];
    
    /** @var Field[] */
    protected ?array $_ref = []; // ref
    
    /** @var Field[] */
    protected ?array $_oor = []; // one to one (fk in ref)
    
    /** @var Field[] */
    protected ?array $_om = []; // one to many
    
    /**
     * Array dinamico para identificar a la entidad en un momento determinado
     * @example ["fecha_anio", "fecha_semestre","persona__numero_documento"]
     * @var string[]
     */
    public array $identifier = [];
    
    /**
     * Valores por defecto para ordenamiento
     * @example ["field1" => "asc", "field2" => "desc", ...]
     * @var array<string, string>
     */
    public array $orderDefault = [];
    
    /**
     * nombres de campos no administrables
     * los campos no insertables son aquellos que reciben se asignan directamente desde el servidor sql
     * @var string[]
     */
    public array $noAdmin = [];
    
    /**
     * Valores principales
     * @var string[]
     */
    public array $main = [];
    
    /**
     * Valores unicos
     * Una entidad puede tener varios campos que determinen un valor único
     * @example field1, field2, field3, ...
     * @var string[]
     */
    public array $unique = [];
    
    /**
     * Valores no nulos
     * @var string[]
     */
    public array $notNull = [];
    
    /**
     * Valores unicos multiples
     * Cada juego de valores unicos multiples se define como una Lista
     * @var string[][]
     */
    public array $uniqueMultiple = [];
    
    /** @var array<string, EntityTree> */
    public array $tree = [];
    
    /** @var array<string, EntityRelation> */
    public array $relations = [];
    
    /** @var array<string, EntityRef> */
    public array $oo = [];
    
    /** @var array<string, EntityRef> */
    public array $om = [];
    
    /**
     * Campo de identificacion
     * Si existe un solo campo pk, entonces la pk sera el id.
     * Si existe al menos un campo unique not null, se toma como id.
     * Si existe multiples campos pk, se toman la concatenacion como id.
     * Si existe multiples campos uniqueMultiple, se toman la concatenacion como id.
     * @var string[]
     */
    public array $id = [];
    
    /** @var array<string, Field> */
    public array $fields = [];
    
    public function getFieldNames(): array
    {
        return array_keys($this->fields);
    }
    
    public function getSchema_(): string
    {
        return empty($this->schema) ? "" : $this->schema;
    }
    
    public function getSchemaName(): string
    {
        return $this->schema . $this->name;
    }
    
    public function getSchemaNameAlias(): string
    {
        return $this->schema . $this->name . " AS " . $this->alias;
    }
    
    /**
     * @param string[] $fieldNames
     * @return Field[]
     */
    protected function _Fields(array $fieldNames): array
    {
        $fields = [];
        foreach ($fieldNames as $fieldName) {
            $fields[] = $this->db->Field($this->name, $fieldName);
        }
        return $fields;
    }
    
    /**
     * fields no fk
     * @return Field[]
     */
    public function Fields(): array
    {
        return $this->_Fields($this->getFieldNames());
    }
    
    /**
     * fields many to one
     * @return Field[]
     */
    public function FieldsFk(): array
    {
        return $this->_Fields($this->fk);
    }
    
    /**
     * @return Field[]
     */
    public function FieldsRef(): array
    {
        if ($this->_ref !== null) {
            return $this->_ref;
        }
        
        $this->_ref = [];
        $this->_oor = [];
        $this->_om = [];
        
        foreach ($this->db->entities as $entityName => $entity) {
            foreach ($entity->fk as $fieldName) {
                $field = $this->db->Field($entityName, $fieldName);
                if ($field->refEntityName === $this->name) {
                    $this->_ref[] = $field;
                    if (in_array($field->name, $entity->unique)) {
                        $this->_oor[] = $field;
                    } else {
                        $this->_om[] = $field;
                    }
                }
            }
        }
        
        return $this->_ref;
    }
    
    /**
     * @return Field[]
     */
    public function FieldsOor(): array
    {
        if ($this->_oor !== null) {
            return $this->_oor;
        }
        
        $this->FieldsRef();
        return $this->_oor;
    }
    
    /**
     * @return Field[]
     */
    public function FieldsOm(): array
    {
        if ($this->_om !== null) {
            return $this->_om;
        }
        
        $this->FieldsRef();
        return $this->_om;
    }
    
    public function Pf(?string $fieldId = null): string
    {
        return !empty($fieldId) ? $fieldId . $this->db->config->separator : "";
    }
    
    public function Pt(?string $fieldId = null): string
    {
        return !empty($fieldId) ? $fieldId : $this->alias;
    }
    
    public function Map(string $fieldName, ?string $fieldId = null): string
    {
        return $this->Pt($fieldId) . "." . $fieldName;
    }
    
    public function CleanPf(string $fieldName): string
    {
        $pf = $this->Pf();
        if (!empty($pf) && str_contains($fieldName, $pf)) {
            return str_replace($pf, "", $fieldName);
        }
        return $fieldName;
    }
}